subsys/mctp/mctp_uart.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/mctp/mctp_uart.h>
+#include <crc-16-ccitt.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mctp_uart, CONFIG_MCTP_LOG_LEVEL);
+
+#define MCTP_UART_REVISION     0x01
+#define MCTP_UART_FRAMING_FLAG 0x7e
+#define MCTP_UART_ESCAPE       0x7d
+
+const char *UART_EVENT_STRING[] = {
+	"TX Done",     "TX Aborted", "RX Ready", "RX Buffer Request", "RX Buffer Released",
+	"RX Disabled", "RX Stopped",
+};
+
+const char *MCTP_STATE_STRING[] = {
+	"Wait: Sync Start", "Wait: Revision", "Wait: Len",  "Data",
+	"Data: Escaped",    "Wait: FCS1",     "Wait: FCS2", "Wait: Sync End",
+};
+
+struct mctp_serial_header {
+	uint8_t flag;
+	uint8_t revision;
+	uint8_t len;
+};
+
+struct mctp_serial_trailer {
+	uint8_t fcs_msb;
+	uint8_t fcs_lsb;
+	uint8_t flag;
+};
+
+static inline struct mctp_binding_uart *binding_to_uart(struct mctp_binding *b)
+{
+	return (struct mctp_binding_uart *)b;
+}
+
+static void mctp_uart_finish_pkt(struct mctp_binding_uart *uart, bool valid)
+{
+	struct mctp_pktbuf *pkt = uart->rx_pkt;
+
+	__ASSERT_NO_MSG(pkt);
+
+	if (valid) {
+		mctp_bus_rx(&uart->binding, pkt);
+	}
+
+	uart->rx_pkt = NULL;
+}
+
+static void mctp_uart_start_pkt(struct mctp_binding_uart *uart, uint8_t len)
+{
+	uart->rx_pkt = mctp_pktbuf_alloc(&uart->binding, len);
```

Comment:
A common practice I see often is developers re-using variables.  In this case would it make sense to ASSERT on entering if the `uart->rx_pkt` is not NULL?  I believe the whole `mctp_pktbuf_alloc` does actually do an `m_alloc` or `m_realloc` which could result in a memory leak here.

Reply:
Yes it should be NULL when allocating, https://github.com/zephyrproject-rtos/zephyr/pull/75743/files#diff-c9d952605a6e4b48ea6b1e57729eb4bd6ab93f3f3890976b73b04fe43ac3ac60R59 NULLs the pointer once its moved into libmctp's call (which eventually I believe leads to the application callbacks and such).